streams/dlna_meta.py

```diff
@@ -1,68 +1,210 @@
+    and a directory to place cover art in.
+    a fifo will be created at the path provided if it does not exist.
 """
-import sys
-import re
+
 import argparse
+import upnpy
+from upnpy.ssdp.SSDPDevice import SSDPDevice
+import json
 import os
+import sys
+import threading
+import time
+import xml.etree.ElementTree as ET
 
-# parse the log file generated by gmrender-resurrect, extracting out the track info
+# the interval at which to update the metadata file
+METADATA_UPDATE_INTERVAL = 3 # seconds
 
-# TODO: dlna generates state events to subscribers, in the future we should use this method instead
-# Check out https://github.com/hzeller/upnp-display/blob/master/controller-state.cc for details
+MIN_STOP_TIME = 10 # seconds
 
-parser = argparse.ArgumentParser(description="dlna metadata interpreter")
-parser.add_argument('dlna_config_dir')#, description='directory used for dlna song info generation')
-args = parser.parse_args()
+# Reads metadata from the upnp AVTransport and writes it to a json file in a loop.
+def metadata_reader(metadata_path: str, album_art_dir: str, service: SSDPDevice, debug: bool = False):
+  """ Read metadata from the upnp AVTransport and write it to a file. """
+  last_file = ''           # last cover art file downloaded
+  stop_counter = 0         # counter to prevent empty metadata on transitions
+  metadata = {"state":"",  # metadata dict to write to file
+              "title":"",
+              "artist":"",
+              "album":"",
+              "album_art":""}
+
+  with open(metadata_path, 'w') as f:
+    while True:
+
+      # sleep for a bit
+      time.sleep(METADATA_UPDATE_INTERVAL)
+
+      try:
+
+        # try to get the transport state and convert it to a valid string
+        try:
+          transport_state = service.GetTransportInfo(InstanceID=0)["CurrentTransportState"]
+          if(transport_state == "PLAYING"):
+            metadata["state"] = "playing"
+          elif(transport_state == "PAUSED_PLAYBACK"):
+            metadata["state"] = "paused"
+          elif(transport_state == "STOPPED"):
+            stop_counter += METADATA_UPDATE_INTERVAL
+            if stop_counter >= MIN_STOP_TIME: # only set to stopped if it's been stopped for a few seconds
+              metadata["state"] = "stopped"
+          else:
+            metadata["state"] = "playing" # default to playing for states like TRANSITIONING
+
+          if transport_state == "TRANSITIONING": # metadata is empty while transitioning so just skip this state
+            continue
+
+          if transport_state != "STOPPED": # reset the stop counter if the state is not stopped
+            stop_counter = 0
+
+        except Exception as e:
+          print(f"Error: could not get transport state: {e}")
+
+        # try to get song-info from the service and parse it
+        try:
+          meta_xml = ET.fromstring(service.GetMediaInfo(InstanceID=0)["CurrentURIMetaData"])
+          for i in meta_xml.iter():
```

Comment:
We may want to be careful here about synchronizing artist/title/album/album_art data. Should all be set to ''  or None when we get new metadata?